### PR TITLE
Changes for removing deadcode in taint_tolerations

### LINKED ISSUE #43007: [minor]  Dead code in taints_tolerations
Kubernetes version (use kubectl version):

master

What happened:

Unnecessary always nil return value for function getTolerationListFromPod.

What you expected to happen:

We need to remove error return value for getTolerationListFromPod function.

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #43007

### CODE REVIEW

**Comment by jayunit100** [comment]
yeah this seems sensible,  post 1.7 obviously .

**Comment by ravisantoshgudimetla** [comment]
@kubernetes/sig-scheduling-pr-reviews

**Comment by jayunit100** [comment]
cc @kubernetes/sig-scheduling-pr-reviews

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go [hunk 1]
```diff
@@ -52,8 +52,8 @@ func getAllTolerationPreferNoSchedule(tolerations []v1.Toleration) (tolerationLi
 	return
 }
 
-func getTolerationListFromPod(pod *v1.Pod) ([]v1.Toleration, error) {
```
Comment (aveshagarwal) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: remove getTolerationListFromPod and just have getAllTolerationPreferNoSchedule, it seems unnecessary/confusing to wrap a specific function for PreferNoSchedule.

Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: I agree with Avesh. The fact that this wrapper drops "PreferNoSchedule" from the name of the wrapped function gives the impression to the reader that the function returns all of the pod tolerations.

Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: Makes sense @aveshagarwal and @bsalamat - Initially I wanted to do the same but since the issue was related to eliminating deadcode, I didnot touch that. If you guys are ok with including the change, I can do that here itself.

Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: Sure, please go ahead and do so. After dropping `error` from the return values, there is no reason to keep this method. It only confuses people.

Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: @aveshagarwal @bsalamat, I have updated the code as per your suggestions. I also modified some parts of code as per go standards. Please let me know if any other changes are needed.

**Comment by jayunit100** [comment]
@k8s-bot test this please

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go [hunk 2]
```diff
@@ -204,6 +204,18 @@ func TestTaintAndToleration(t *testing.T) {
 				{Host: "nodeC", Score: 0},
 			},
 		},
+		{
+			test: "Default behaviour No taints and tolerations, lands on node with no taints",
```
Comment (jayunit100) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go]: thanks for adding this as well ; clarifies whats going on here.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go [hunk 3]
```diff
@@ -43,40 +42,33 @@ func countIntolerableTaintsPreferNoSchedule(taints []v1.Taint, tolerations []v1.
 
 // getAllTolerationEffectPreferNoSchedule gets the list of all Toleration with Effect PreferNoSchedule
 func getAllTolerationPreferNoSchedule(tolerations []v1.Toleration) (tolerationList []v1.Toleration) {
-	for i := range tolerations {
-		toleration := &tolerations[i]
-		if len(toleration.Effect) == 0 || toleration.Effect == v1.TaintEffectPreferNoSchedule {
```
Comment (aveshagarwal) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: why did you remove len(toleration.Effect) == 0? please put it back.

Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: I thought we don't need it as we are comparing with constants, so if it is needed probably && would have helped as for short circuit evaluation.

Comment (aveshagarwal) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: No, empty effect maps to all effects that means PreferNoSchedule too.

Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: Ok, I can add that.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go [hunk 4]
```diff
@@ -43,40 +42,33 @@ func countIntolerableTaintsPreferNoSchedule(taints []v1.Taint, tolerations []v1.
...
-		tolerationList = priorityMeta.podTolerations
+		tolerationsPreferNoSchedule = priorityMeta.podTolerations
+
 	} else {
-		var err error
-		tolerationList, err = getTolerationListFromPod(pod)
-		if err != nil {
-			return schedulerapi.HostPriority{}, err
-		}
+		tolerations := pod.Spec.Tolerations
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: nit: similar to the above, please remove `tolerations`.

**Comment by ravisantoshgudimetla** [comment]
Made the changes @aveshagarwal @bsalamat.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go [hunk 5]
```diff
@@ -27,56 +27,47 @@ import (
...
 	return
 }
 
-// getAllTolerationEffectPreferNoSchedule gets the list of all Toleration with Effect PreferNoSchedule
+// getAllTolerationEffectPreferNoSchedule gets the list of all Tolerations with Effect PreferNoSchedule or with no effect.
 func getAllTolerationPreferNoSchedule(tolerations []v1.Toleration) (tolerationList []v1.Toleration) {
-	for i := range tolerations {
-		toleration := &tolerations[i]
+	for _, toleration := range tolerations {
 		if len(toleration.Effect) == 0 || toleration.Effect == v1.TaintEffectPreferNoSchedule {
```
Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: In order to avoid confusion in the future and now that you are at it, could you please add a comment before this line:
// Empty effect means all effects which includes PreferNoSchedule.

Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: This is not for this PR, but we should add a test to cover empty effect (if we don't have already) in a separate PR.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go [hunk 6]
```diff
@@ -204,6 +204,18 @@ func TestTaintAndToleration(t *testing.T) {
...
+			//pod without tolerations
+			pod: podWithTolerations([]v1.Toleration{}),
+			nodes: []*v1.Node{
+				//Node without taints
```
Comment (aveshagarwal) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go]: I think add one more node with some taints. And then test to see if the pod lands on the node without taints. So just adding one more node with taints should do.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go (hunk 5 above)
Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: Makes sense, I added one more line at the top of function which tells that this function can collect all tolerations that have either no effect or PreferNoSchedule.

I added a test that has both empty toleration and taint. I will add one that has empty toleration and a taint separately.

Comment (bsalamat) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: Yeah, but it is not clear from that comment why we are collecting tolerations with empty effect.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go (hunk 6 above)
Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go]: I am slightly confused, are you telling that the node with taints should get a score of 0? It should not get 0 unless the effect is TaintEffectPreferNoSchedule? - Correct me if I am wrong and that case has been handled already.

Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go]: I added one with TaintEffectPreferNoSchedule. Let me know, if this is the case that you want.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go (hunk 5 above)
Comment (ravisantoshgudimetla) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration.go]: Added.

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go (hunk 6 above)
Comment (jayunit100) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go]: Yeah i think that reflects what avesh was talking about.

**Comment by jayunit100** [comment]
@k8s-bot test this please

File: plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go (hunk 6 above)
Comment (aveshagarwal) [inline plugin/pkg/scheduler/algorithm/priorities/taint_toleration_test.go]: yeah right.

**Comment by aveshagarwal** [comment]
LGTM.

**Comment by bsalamat** [comment]
Thanks, @ravisantoshgudimetla!

**Comment by ravisantoshgudimetla** [comment]
@bsalamat, @aveshagarwal - Can you guys add a release note for 1.7 now that 1.7 patches could be merged?

**Comment by ravisantoshgudimetla** [comment]
Accidentally closed and re-opened it..

**Comment by bsalamat** [comment]
@ravisantoshgudimetla I think we don't need a release note for this PR as it is not making any change to the behavior of the code.
